XR_APILAYER_NOVENDOR_toolkit/fsr.cpp

```diff
@@ -0,0 +1,216 @@
+
+                // implementation is using a shader compilation define for this
+                //config.Const4[3] = (hdr != NISHDRMode::None) ? 1 : 0;
+    
+                m_configBuffer = m_device->createBuffer(sizeof(config), "FSR Constants CB", &config, true);
+            }
+        }
+
+        void upscale(std::shared_ptr<ITexture> input, std::shared_ptr<ITexture> output, int32_t slice = -1) override {
+            if (!m_intermediary) {
+                const auto& infos = output->getInfo();
+                initializeIntermediary(infos.width, infos.height, 0 /* infos.format */);
+            }
+
+            if (!m_isSharpenOnly) {
+                m_device->setShader(m_shaderEASU);
+                m_device->setShaderInput(0, m_configBuffer);
+                m_device->setShaderInput(0, input, slice);
+                m_device->setShaderOutput(0, m_intermediary);
+                m_device->dispatchShader();
+            }
+
+            m_device->setShader(m_shaderRCAS);
+            m_device->setShaderInput(0, m_configBuffer);
+            m_device->setShaderInput(0, m_isSharpenOnly ? input : m_intermediary);
+            m_device->setShaderOutput(0, output, slice);
+            m_device->dispatchShader();
+        }
+
+      private:
+        void initializeScaler() {
+            const auto shadersDir = std::filesystem::path(dllHome) / std::filesystem::path("shaders");
+            const auto shaderPath = shadersDir / std::filesystem::path("FSR.hlsl");
+
+            // This value is the image region dimension that each thread group of the FSR shader operates on
+            const auto threadGroupWorkRegionDim = 16u;
+            const std::array<unsigned int, 3> threadGroups = {
+                (m_outputWidth  + (threadGroupWorkRegionDim - 1)) / threadGroupWorkRegionDim, // dispatchX
+                (m_outputHeight + (threadGroupWorkRegionDim - 1)) / threadGroupWorkRegionDim, // dispatchY
+                1 
+            };
+
+            // EASU/RCAS common
+            utilities::shader::Defines defines;
+            defines.add("FSR_THREAD_GROUP_SIZE", 64);
+            defines.add("SAMPLE_SLOW_FALLBACK", 1);
+            defines.add("SAMPLE_BILINEAR", 0);
+            defines.add("SAMPLE_HDR_OUTPUT", 0);
+
+            // EASU specific
+            defines.add("SAMPLE_RCAS", 0);
+            defines.add("SAMPLE_EASU", 1);
+            m_shaderEASU = m_device->createComputeShader(
+                shaderPath.string(), "mainCS", "FSR EASU CS", threadGroups, defines.get(), shadersDir.string());
+
+            // RCAS specific
+            defines.set("SAMPLE_EASU", 0);
+            defines.set("SAMPLE_RCAS", 1);
+            m_shaderRCAS = m_device->createComputeShader(
+                shaderPath.string(), "mainCS", "FSR RCAS CS", threadGroups, defines.get(), shadersDir.string());
+
+            m_isSharpenOnly = false;
+        }
+
+        void initializeSharpen() {
+            initializeScaler();
+
+            // TODO
```

Comment:
Will this be a future optimization?

Reply:
Looking at the `upscale()` method, the case of sharpening only looks implemented?

Reply:
Yes it is left open for future enhancement. AMD offers 2 libs: FSR and RCAS. The former does both, the latter seems to be designed for the sharpen only case.

Reply:
Can you copy this comment inside the TODO so it's clear what's left to do?